src/OrleansSQLUtils/Storage/Provider/RelationalStorageProvider.cs

```diff
@@ -0,0 +1,550 @@
+        RelationalProviderNoDeserializer = RelationalProviderBase + 10,
+        RelationalProviderNoStateFound = RelationalProviderBase + 11,
+        RelationalProviderClearing = RelationalProviderBase + 12,
+        RelationalProviderCleared = RelationalProviderBase + 13,
+        RelationalProviderReading = RelationalProviderBase + 14,
+        RelationalProviderRead = RelationalProviderBase + 15,
+        RelationalProviderReadError = RelationalProviderBase + 16,
+        RelationalProviderWriting = RelationalProviderBase + 17,
+        RelationalProviderWrote = RelationalProviderBase + 18,
+        RelationalProviderWriteError = RelationalProviderBase + 19
+    }
+
+
+    /// <summary>
+    /// A storage provider for writing grain state data to relational storage.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Required configuration params: <c>DataConnectionString</c>
+    /// </para>
+    /// <para>
+    /// Optional configuration params:
+    /// <c>AdoInvariant</c> -- defaults to <c>System.Data.SqlClient</c>
+    /// <c>UseJsonFormat</c> -- defaults to <c>false</c>
+    /// <c>UseXmlFormat</c> -- defaults to <c>false</c>
+    /// <c>UseBinaryFormat</c> -- defaults to <c>true</c>
+    /// </para>
+    /// </remarks>
+    [DebuggerDisplay("Name = {Name}, ConnectionString = {Storage.ConnectionString}")]
+    public class RelationalStorageProvider: IStorageProvider
+    {
+        /// <summary>
+        /// The canonical constant for storage provider connection string. Corresponds to Orleans.Runtime.Constants.DATA_CONNECTION_STRING_NAME.
+        /// </summary>
+        private const string DataConnectionStringPropertyName = "DataConnectionString";
+
+        /// <summary>
+        /// The canonical constant for relational storage provider connection string. Corresponds to Orleans.Runtime.Constants.ADO_INVARIANT_NAME.
+        /// </summary>
+        private const string DataConnectionInvariantPropertyName = "AdoInvariant";
+
+        /// <summary>
+        /// The Service ID for which this relational provider is used.
+        /// </summary>
+        private string ServiceId { get; set; }
+
+        /// <summary>
+        /// The storage used for back-end operations.
+        /// </summary>
+        private IRelationalStorage Storage { get; set; }
+
+        /// <summary>
+        /// The default query to initialize this structure from the Orleans database.
+        /// </summary>
+        public const string DefaultInitializationQuery = "SELECT QueryKey, QueryText FROM OrleansQuery WHERE QueryKey = 'WriteToStorageKey' OR QueryKey = 'ReadFromStorageKey' OR QueryKey = 'ClearStorageKey';";
+
+        /// <summary>
+        /// The default ADO.NET invariant used for storage if none is given. This corresponds to Orleans.Runtime.Constants.INVARIANT_NAME_SQL_SERVER.
+        /// </summary>
+        public const string DefaultAdoInvariantInvariantPropertyName = "System.Data.SqlClient";
```

Comment:
Very true that about using the constant. I was using the suffix `PropertyName` as it was used in other places too. I also a few lines about it to the lengthy introduction that if we should standardize them between providers and even let all interfaces provide a list of the supported properties with a short explanation. This way one could search by `IStorageProvider` interface and dumb supported properties.